### PR TITLE
Remove most special handling of MKL in CMake configuration

### PR DESCRIPTION
**Pull request by msimberg**
- From version 0.5 onwards the spack package uses spack's blas/lapack libraries
- The `DLAF_WITH_MKL_LEGACY` option is completely removed
- MKL is auto-detected

This is currently untested, but early feedback more than welcome and the general structure is in place!

The first version of the MKL auto-detection simply uses `try_compile` with `LAPACK_LIBRARY` and a call to `mkl_set_num_threads_local`. It uses the result of this `try_compile` as the default value of `DLAF_WITH_MKL`. It's important to note that currently this will only be tested on the first configuration, so if the first configuration is wrong, and one then changes `LAPACK_LIBRARY` to have the correct paths, `DLAF_WITH_MKL` will still be off.

On the other hand, the above method means that one can simply set `DLAF_WITH_MKL` manually to on if one expects MKL to be there, and compilation will then fail if it's not there (instead of quietly disabling it).

The logic could be changed/improved to have different behaviour in the above case.

An alternative method is to simply use `__has_include` in `single_threaded_blas.cpp`. This doesn't allow overriding and is silent in either case (MKL found or not found).

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

File: CMakeLists.txt
Comment (rasolca) [inline CMakeLists.txt:11]: What is the benefit of bumping to 3.25?

Comment (msimberg) [reply CMakeLists.txt:11]: It just gives access to a newer form of `try_compile` which doesn't require specifying the build directory: https://cmake.org/cmake/help/v3.25/command/try_compile.html. Not a must at all, I just went with this for the first iteration until we decide what to do exactly.

File: ci/.gitlab-ci.yml
Comment (msimberg) [inline ci/.gitlab-ci.yml:2]: To do: uncomment before merging.

File: ci/docker/build.Dockerfile
Comment (msimberg) [inline ci/docker/build.Dockerfile:18]: This is required because with newer versions of spack `openssl` is detected as an external but the header files are not there without `libssl-dev`, leading to build failures later.

File: spack/packages/dla-future/package.py
Comment (msimberg) [inline spack/packages/dla-future/package.py:249]: To do: change based on what we do in CMake configuration.

File: spack/packages/dla-future/package.py
Comment (rasolca) [inline spack/packages/dla-future/package.py]: I will not switch to auto-detection here.

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — COMMENTED** [review]
The include directory is needed, otherwise non-spack builds are not possible.

**Review by albestro — COMMENTED** [review]
LGTM

File: CMakeLists.txt
Comment (RMeli) [reply CMakeLists.txt:11]: Needs an update to the Spack package too if we go forward with this update.

**Comment by msimberg** [comment]
cscs-ci run

File: CMakeLists.txt
Comment (msimberg) [inline CMakeLists.txt:63]: To do.

File: CMakeLists.txt
Comment (msimberg) [inline CMakeLists.txt:67]: Remove.

File: cmake/FindLAPACK.cmake
Comment (msimberg) [inline cmake/FindLAPACK.cmake:69]: Is there any reasonable check we can do on this? I think not, but I may be wrong.

File: cmake/FindSCALAPACK.cmake
Comment (msimberg) [inline cmake/FindSCALAPACK.cmake:35]: Same question here as for `LAPACK_INCLUDE_DIR`.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

File: cmake/FindSCALAPACK.cmake
Comment (rasolca) [reply cmake/FindSCALAPACK.cmake:35]: Not needed, as we don't include `Cblacs.h` or `mkl_blacs.h` directly.

File: cmake/FindLAPACK.cmake
Comment (rasolca) [inline cmake/FindLAPACK.cmake:71]: The only case in which we currently need it is to include the headers for mkl_set_num_threads.

File: CMakeLists.txt
Comment (msimberg) [reply CMakeLists.txt]: I've changed it to use the older form of `try_compile` and reverted the version bump.

File: cmake/FindLAPACK.cmake
Comment (msimberg) [reply cmake/FindLAPACK.cmake:71]: Yep, I removed the TODO. The worst case is that the build will fail if the include directory isn't set properly, instead of failing during configuration time. It's a failure in any case.

File: CMakeLists.txt
Comment (msimberg) [inline CMakeLists.txt:64]: To do: the description is outdated (this option applies to all MKL versions now).

**Comment by msimberg** [comment]
I've probably messed something up in the last pushes, but in theory this is close to how I imagine the final state to be. I've made separable compilation with CUDA conditional on the umpire version, with a comment: https://github.com/eth-cscs/DLA-Future/blob/460d24f4e41a20bb34481dae8a45664fdd755a3d/src/CMakeLists.txt#L61-L69.

Known open issues are:
- if and how to cache the `try_compile` result for detecting MKL
- if and how to report the result of MKL detection

The rest is obviously open to comments as well, but I've attempted to clean it up as far as I can.

**Comment by msimberg** [comment]
I've changed back the CUDA CI to use `-Werror=all-warnings` since we now don't get linker warnings with the newest version of Umpire. We will still get the warnings on builds with older versions of Umpire, but outside of CI those warnings should not be fatal. I opened https://github.com/eth-cscs/DLA-Future/issues/1154 as a reminder to potentially look at the other warnings.

File: CMakeLists.txt
Comment (rasolca) [reply CMakeLists.txt:62]: Something like:
`Intel MKL detection: [Not ]Found`?

Comment (msimberg) [reply CMakeLists.txt:62]: Works for me. `message(STATUS)` is probably sufficient? Or do you think it's useful to make the not found case `message(WARNING)`? I'd worry it's too noisy if one isn't actually interested in using MKL.

**Review by rasolca — APPROVED** [review]
> if and how to cache the try_compile result for detecting MKL

Not sure... maybe @albestro has an idea?

File: CMakeLists.txt
Comment (albestro) [inline CMakeLists.txt:57]:
```suggestion
  DLAF_WITH_MKL_TRY_COMPILE "${PROJECT_BINARY_DIR}/cmake/tests"
```

Otherwise it will end up polluting the source dir (we have plenty of rules in `.gitignore` so we don't clutter up the repo anyway).

File: CMakeLists.txt
Comment (albestro) [reply CMakeLists.txt:62]: I agree with both of you. A **STATUS** message like **"Intel MKL: (NOT) Found"** saying if MKL has been found or not.

**Comment by albestro** [comment]
> > if and how to cache the try_compile result for detecting MKL
> 
> Not sure... maybe @albestro has an idea?

I would say that it might be nicer to not cache, so that any change to `LAPACK_LIBRARY`/`LAPACK_INCLUDE_DIR` will be reflected also in the MKL_TRY_COMPILE result.

Otherwise, let's suppose we do this sequence

- `cmake -DLAPACK_LIBRARY=<working-config> ...`
- `cmake -DLAPACK_LIBRARY=<non-working-config> ...`

with the cached result for MKL we might end up getting BLAS/LAPACK error, because of the `check_function_exists` in the FindLAPACK that explicitly clear the cache, but at the same time we get that MKL is found correctly (because it is checked just at the first time).

So, I would suggest to do the same we do for LAPACK in FindLAPACK, i.e. `unset(DLAF_WITH_MKL_TRY_COMPILE CACHE)`. And, trivially, in the future start using `NO_CACHE` where possible (apparently up to 3.29 it is possible for `try_compile` but not for `check_function_exists`).

File: CMakeLists.txt
Comment (RMeli) [reply CMakeLists.txt:62]: I think a warning would break CP2K CI.

File: CMakeLists.txt
Comment (msimberg) [reply CMakeLists.txt]: Good catch! That was never meant to be in the source directory...

File: CMakeLists.txt
Comment (msimberg) [reply CMakeLists.txt]: Changed it to a status message.

**Comment by msimberg** [comment]
> > > if and how to cache the try_compile result for detecting MKL
> > 
> > 
> > Not sure... maybe @albestro has an idea?
> 
> I would say that it might be nicer to not cache, so that any change to `LAPACK_LIBRARY`/`LAPACK_INCLUDE_DIR` will be reflected also in the MKL_TRY_COMPILE result.
> 
> Otherwise, let's suppose we do this sequence
> 
>     * `cmake -DLAPACK_LIBRARY=<working-config> ...`
> 
>     * `cmake -DLAPACK_LIBRARY=<non-working-config> ...`
> 
> 
> with the cached result for MKL we might end up getting BLAS/LAPACK error, because of the `check_function_exists` in the FindLAPACK that explicitly clear the cache, but at the same time we get that MKL is found correctly (because it is checked just at the first time).
> 
> So, I would suggest to do the same we do for LAPACK in FindLAPACK, i.e. `unset(DLAF_WITH_MKL_TRY_COMPILE CACHE)`. And, trivially, in the future start using `NO_CACHE` where possible (apparently up to 3.29 it is possible for `try_compile` but not for `check_function_exists`).

Currently on this branch, the `try_compile` result is used as the _default_ value for `DLAF_WITH_MKL`. This is done with the idea that `DLAF_WITH_MKL` can be set explicitly by a user to signal that they expect MKL to be used or not used (leading to compilation failures if MKL isn't actually set up correctly). In this case it doesn't make much difference if `DLAF_WITH_MKL_TRY_COMPILE` is a cache variable or not, and if it's reset or not because it'll only be used once to set the default.

I think if we reset and redetect MKL every time then `DLAF_WITH_MKL_TRY_COMPILE` should just be `DLAF_WITH_MKL`. I haven't tried it yet, but in this case I think we can simply let `try_compile` create the variable. If the user has already set it the `try_compile` should be skipped (but as I said, not verified). Then a user can `-UDLAF_WITH_MKL` to redetect it if they don't want to set it explicitly.

As a last option, we `try_compile` into `DLAF_WITH_MKL` and don't cache the variable. This, however, means that the user can't override the option (as far as I can think of...).

**Comment by rasolca** [comment]
Not sure what to say... just realized that auto-detection might have false positive if the mkl include directory is present (e.g. if using mkl fftw).
So the possibility for the user should be available.
I think it should work like this:

`DLAF_WITH_MKL`: is a cached value set by user

| DLAF_WITH_MKL | try_compile | -DDLAF_WITH_MKL |
| --- | --- | --- |
| ON | not run | added |
| OFF | not run | skipped |
| Not defined | OK | added |
| Not defined | FAIL | skipped |

If a variable is also needed in the rest of the cmake scripts I would use a different name.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
> I think it should work like this:
> 
> `DLAF_WITH_MKL`: is a cached value set by user
> DLAF_WITH_MKL 	try_compile 	-DDLAF_WITH_MKL
> ON 	not run 	added
> OFF 	not run 	skipped
> Not defined 	OK 	added
> Not defined 	FAIL 	skipped
> 
> If a variable is also needed in the rest of the cmake scripts I would use a different name.

This sounds pretty reasonable. I made an attempt at implementing this logic in https://github.com/eth-cscs/DLA-Future/pull/1149/commits/d5a7ae4624baf4fc66ae63b02fcc461d38ea3acf.

> jst realized that auto-detection might have false positive if the mkl include directory is present (e.g. if using mkl fftw).

Yeah, I was worried about something like this as well. Without intending to I think this now covers that quite reasonably: https://github.com/eth-cscs/DLA-Future/blob/d5a7ae4624baf4fc66ae63b02fcc461d38ea3acf/CMakeLists.txt#L81-L84. When `DLAF_WITH_MKL` is explicitly enabled we can safely disable OpenMP. If `DLAF_WITH_MKL` is undefined that first branch is always false so we'll keep OpenMP enabled (if it wasn't changed by the user). In the worst case we end up calling both `omp_set_num_threads` and `mkl_set_num_threads`, but we won't end up in the situation where we think we're using MKL, but we're not setting `omp_set_num_threads`.

In the spack package we set `DLAF_WITH_MKL` explicitly so there we only end up with one or the other, never both, enabled.

At least this is how I think it's working now and I think it seems reasonable, but not 100% sure.

I think warning/status messages can still be tweaked, but I'll leave that up to you to comment about.

**Comment by msimberg** [comment]
cscs-ci run

File: CMakeLists.txt
Comment (rasolca) [inline CMakeLists.txt:80]:
```suggestion
if(DLAF_WITH_MKL_INTERNAL)
```
?

**Review by rasolca — APPROVED** [review]

File: CMakeLists.txt
Comment (rasolca) [reply CMakeLists.txt:80]: Right better calling twice than never.

File: ci/common-ci.yml
Comment (RMeli) [inline ci/common-ci.yml:37]:
```suggestion
    SPACK_SHA: 1184de8352281ce1b34a8f3f6b54d7fadc2b216a
```
Worth testing with https://github.com/spack/spack/pull/44242?

**Review by RMeli — APPROVED** [review]
LGTM, thanks!

File: ci/common-ci.yml
Comment (msimberg) [reply ci/common-ci.yml:37]: I think it's worth testing, but I'll do that in a follow-up PR.

Comment (rasolca) [reply ci/common-ci.yml:37]: I would postpone this check after the release in a new PR.
If the new one-api package doesn't work, it need to be fixed in spack.

File: CMakeLists.txt
Comment (albestro) [inline CMakeLists.txt:56]: since `DLAF_WITH_MKL_TRY_COMPILE` will be set later, doesn't this mean that by default `DLAF_WITH_MKL=""`?

File: cmake/template/DLAFConfig.cmake.in
Comment (albestro) [inline cmake/template/DLAFConfig.cmake.in:23]: shouldn't `DLAF_WITH_MKL` changed to `DLAF_WITH_MKL_INTERNAL`?

**Review by albestro — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (msimberg) [reply CMakeLists.txt:56]: Hrm, excellent catch. This should not be an option at all, otherwise it won't be undefined if a user doesn't set it. Let me fix that.

File: cmake/FindLAPACK.cmake
Comment (msimberg) [reply cmake/FindLAPACK.cmake]: No, I don't think so.

> False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the **empty string**, or ends in the suffix -NOTFOUND.

It's just a copy-paste of the logic for `LAPACK_LIBRARY`. I changed both in https://github.com/eth-cscs/DLA-Future/pull/1149/commits/1758a6694729dc0a5c12e3b0b94eba4df54c1d04.

File: CMakeLists.txt
Comment (msimberg) [reply CMakeLists.txt]: https://github.com/eth-cscs/DLA-Future/pull/1149/commits/9dc2f78d8ff707e8d6c64f517f5929bc83dcdf63

File: cmake/template/DLAFConfig.cmake.in
Comment (msimberg) [reply cmake/template/DLAFConfig.cmake.in:23]: Or not be set at all? Given that it has `INTERNAL` in the name I don't think we should expect users to rely on `DLAF_WITH_MKL_INTERNAL` to decide anything. We don't need it in `DLAFConfig.cmake` anymore.

If anything stays perhaps `set(DLAF_WITH_MKL @DLAF_WITH_MKL_INTERNAL@)` as that's consistent with the preprocessor definition https://github.com/eth-cscs/DLA-Future/blob/1758a6694729dc0a5c12e3b0b94eba4df54c1d04/src/CMakeLists.txt#L106.

What do you think?

Comment (albestro) [reply cmake/template/DLAFConfig.cmake.in:23]: Good point! it shouldn't be needed anymore.

Comment (msimberg) [reply cmake/template/DLAFConfig.cmake.in:23]: https://github.com/eth-cscs/DLA-Future/pull/1149/commits/5f5bb92d996ec30505f07de56936c21d2330e35b

**Comment by msimberg** [comment]
cscs-ci run

File: ci/common-ci.yml
Comment (msimberg) [reply ci/common-ci.yml]: Since there were still other changes to be made, I also bumped the spack commit on this PR.

**Review by albestro — APPROVED** [review]

File: ci/common-ci.yml
Comment (msimberg) [reply ci/common-ci.yml]: The new commit also works well. On the old commit the gfortran mkl was linked, on the new commit the intel one, so all good there!

**Pull request merged**
